editoast/src/documents.rs

```diff
@@ -0,0 +1,267 @@
+//! This module manage documents in the database.
+//! You can add, retrieve and delete documents.
+//!
+//! Each document has a unique key that you have to provide or build.
+//!
+//! ## Example
+//!
+//! ```rust
+//! let study_pk = 42;
+//! let key = DocumentKey::new_hash(&format!("study_image_{study_pk}"));
+//! let image: Vec<u8> = ...;
+//! let doc = Document::insert(db_pool, &key, ContentType::png(), image).await?;
+//! let document_pk = doc.get_pk();
+//! let image_url = ket.get_url("http://localhost:8090");
+//! Document::delete(db_pool, &key).await?;
+//! ```
+
+use crate::error::{EditoastError, Result};
+use crate::tables::osrd_infra_document;
+use crate::DbPool;
+use actix_web::http::header::ContentType;
+use actix_web::web::{block, Data};
+use diesel::expression_methods::ExpressionMethods;
+use diesel::result::{DatabaseErrorKind, Error as DieselError};
+use diesel::{insert_into, QueryDsl, RunQueryDsl};
+use editoast_derive::EditoastError;
+use sha1::{Digest, Sha1};
+use thiserror::Error;
+
+#[derive(Debug, Error, EditoastError)]
+#[editoast_error(base_id = "documents")]
+enum DocumentErrors {
+    #[error("Document not found")]
+    #[editoast_error(status = 404)]
+    NotFound,
+    #[error("A document with the same key already exists")]
+    #[editoast_error(status = 500)]
+    AlreadyExists,
+}
+
+#[derive(Debug, Insertable)]
+#[diesel(table_name = osrd_infra_document)]
+struct DocumentForm {
+    key: String,
+    content_type: String,
+    data: Vec<u8>,
+}
+
+/// Represent a document key.
+/// Can be created manually or build from hashable objects.
+///
+/// **WARNING**: The key of a document must be unique.
+///
+/// ## Example
+///
+/// Let's say we have an image of a study to save in the database.
+/// A way to construct your unique key can be to use a tuple of the study id and a static `str`.
+///
+/// ```
+/// let study_pk = 42;
+/// let key = DocumentKey::new_hash(&format("study_image_{study_pk}"));
```

Comment:
can't we automatically hash the content and use that as a key? (content-addressing)
if that is not the goal, then what is the key hash used for?

Reply:
In some cases, we don't want to do that. For example for project images. We will probably have a foreign key to the document so when a project is deleted its image is deleted in cascade. This doesn't work if the key used is the hash of the image (when two projects have the same image).

Using the hash of the document as the key will probably be used for liveries.